### PR TITLE
issue/214-es6-upgrade

### PR DESCRIPTION
**Pull request by littlespex**
This PR updates the code base and build tools to ES6. It replaces jshint, browserify and closure compiler with aslant, rollup, terser and typescript (only for ES5 compilation and basic type declarations, the source files are still JS). If a full typescript conversion is desired, I can piggy back that change onto this PR.

### CODE REVIEW

**Comment by littlespex** [comment]
@palemieux Here is base ES6 conversion. Let me know if you'd be interested in converting to TypeScript.

**Comment by palemieux** [comment]
> Here is base ES6 conversion. Let me know if you'd be interested in converting to TypeScript.

Thanks @littlespex ! Will review.

File: Gruntfile.cjs [hunk 1]
```diff
@@ -43,71 +43,42 @@ module.exports = function (grunt) {
...
+            compile: {
+                cmd: 'tsc',
+            },
+
+            lint: {
+                cmd: 'eslint src',
+            },
+
+            bundle: {
+                cmd: 'rollup -c rollup.config.js',
```
Comment (palemieux) [inline Gruntfile.cjs]:
```suggestion
                cmd: 'npx rollup -c rollup.config.js',
```

File: Gruntfile.cjs [hunk 2]
```diff
@@ -43,71 +43,42 @@ module.exports = function (grunt) {
...
-                ].join("&&")
-            }
-        }
-    }
+            compile: {
+                cmd: 'tsc',
+            },
+
+            lint: {
+                cmd: 'eslint src',
```
Comment (palemieux) [inline Gruntfile.cjs]:
```suggestion
                cmd: 'npx eslint src',
```

File: eslint.config.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      globals: {
+        ...globals.browser,
+        QUnit: true,
+        tests: true,
+      },
+    },
+    rules: {
+      'no-var': 'error',
+      'prefer-const': 'error',
+      'quotes': ['error', 'single', { 'avoidEscape': true, 'allowTemplateLiterals': true }],
```
Comment (palemieux) [inline eslint.config.js]:
```suggestion
      'quotes': ['error', 'double', { 'avoidEscape': true, 'allowTemplateLiterals': true }],
```

For consistency with the current codebase.

**Review by palemieux — CHANGES_REQUESTED** [review]
I am getting the following error when running `grunt build`:

`>> [!] RollupError: dist/main/styles.js (27:37): "parseFloat" is not exported by "dist/main/utils.js", imported by "dist/main/styles.js".`

File: package.json [hunk 4]
```diff
@@ -24,28 +25,37 @@
...
     "author": "Pierre-Anthony Lemieux <pal@sandflow.com>",
-    "main": "src/main/js/main.js",
+    "main": "dist/main/js/main.js",
+    "types": "dist/main/js/main.d.ts",
     "unpkg": "dist/imsc.min.js",
     "scripts": {
-        "prepublishOnly": "grunt build:release"
+        "prepublishOnly": "grunt build:release",
+        "dev": "npx http-server build/public_html",
+        "fix": "eslint --fix src"
```
Comment (palemieux) [inline package.json]:
```suggestion
        "fix": "npx eslint --fix src"
```

File: Gruntfile.cjs [hunk 5]
```diff
@@ -43,71 +43,42 @@ module.exports = function (grunt) {
...
-            {
-                cmd: [
-                    "npx google-closure-compiler --js=<%= properties.umdBuildDir %>/<%= properties.umdAllDebugName %> --js_output_file=<%= properties.umdBuildDir %>/<%= properties.umdAllMinName %>",
-                    "npx google-closure-compiler --js=<%= properties.umdBuildDir %>/<%= properties.umdDebugName %> --js_output_file=<%= properties.umdBuildDir %>/<%= properties.umdMinName %>"
-                ].join("&&")
-            }
-        }
-    }
+            compile: {
+                cmd: 'tsc',
```
Comment (palemieux) [inline Gruntfile.cjs]:
```suggestion
                cmd: 'npx tsc',
```

**Review by palemieux — CHANGES_REQUESTED** [review]

**Comment by littlespex** [comment]
Fixed the bad import and reformatted with the lint changes.

**Comment by palemieux** [comment]
@littlespex Still seeing:

`>> [!] RollupError: dist/main/styles.js (27:37): "parseInt" is not exported by "dist/main/utils.js", imported by "dist/main/styles.js".`

Do you know why I would be seeing it and not you?

**Review by palemieux — CHANGES_REQUESTED** [review]
Still cannot build but getting closer :)

**Comment by littlespex** [comment]
> @littlespex Still seeing:
> 
> `>> [!] RollupError: dist/main/styles.js (27:37): "parseInt" is not exported by "dist/main/utils.js", imported by "dist/main/styles.js".`
> 
> Do you know why I would be seeing it and not you?

Because I had local changes that weren't committed :)

**Comment by palemieux** [comment]
@littlespex An initial objective of imscJS was to allow  `generateISD()` to be called from node since it did not contain any DOM dependencies. Would anything in the refactoring prevent this?

**Comment by littlespex** [comment]
The only difference is that it would need to be imported now instead of required. I can update the build step to generate a cjs version of the library as well if you want to keep a "requirable" version of the lib.

**Comment by palemieux** [comment]
> I can update the build step to generate a cjs version of the library as well if you want to keep a "requirable" version of the lib.

Let's definitely not support "require" :)

**Comment by palemieux** [comment]
@littlespex I am thinking we could create a 1.2 branch using yours as a basis... we might have other improvements, e.g., factoring out XML parsing.

**Comment by littlespex** [comment]
@palemieux I've added a GitHub action to run CI (lint, build, and test) for every pull request. The unit tests in `src/test/js` and `src/test/webapp/js/unit-test.js` have been converted to use node's built in test runner. The original tests have been left in place, but doesn't appear that any of them actually need to be run in a browser environment. Can we remove them in favor of the new ones?

**Comment by palemieux** [comment]
> Can we remove them in favor of the new ones?

+1

**Comment by palemieux** [comment]
@littlespex Ok to branch or do you have more improvements to make re: conversion to ES6 + modern tooling

**Comment by littlespex** [comment]
I don't have anything else.

**Comment by palemieux** [comment]
@littlespex One last request/question: can we move the `./test/*.js` under `./src/test/js` and `./test/resources/*` under `./src/test/resources/`?

**Comment by littlespex** [comment]
Done

**Comment by palemieux** [comment]
```
>> dist/main/main.js → dist/imsc.all.debug.js, dist/imsc.all.min.js...
>> (!) Circular dependencies
>> polyfill-node._stream_duplex.js -> polyfill-node._stream_readable.js -> polyfill-node._stream_duplex.js
>> polyfill-node._stream_duplex.js -> polyfill-node._stream_writable.js -> polyfill-node._stream_duplex.js
>> created dist/imsc.all.debug.js, dist/imsc.all.min.js in 1s
```

Is this an issue?

**Comment by littlespex** [comment]
No. It's just a warning when rollup bundles the `sax` package. It resolves these correctly and bundles the "all" version of the library without error.

**Comment by palemieux** [comment]
Created https://github.com/sandflow/imscJS/tree/integration/v2 and https://github.com/sandflow/imscJS/milestone/12

Thanks!

**Pull request closed without merging**